### PR TITLE
far2l: 2019-12-14 -> 2020-12-30 

### PR DESCRIPTION
**Pull request by hypersw**
###### Motivation for this change

`far2l` were not able to run on a GUI-less NixOS installation, just exited silently. Upon a short diagnostic together with folks from far2l, ruled out that we had a version bundled which didn't have the tty-compat code yet.

###### Things done

Updated to a fresh `master` revision, fixed compilation, checked that it runs and appears operational.
The app build has some packages optional, needed to enable operations of plugins like network share browsers, added them to the build.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)

- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) -- found no relevant tests
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` -- the command messed up in patches, but I believe no packages depend on far2l, based on repo search
- [x] Tested execution of all binary files (usually in `./result/bin/`) -- tested running some of the plugins, incl which were just added
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date -- I think there're no docs outside package
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hypersw** [comment]
> `anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/applications/misc/far2l/default.nix:1:1 called without required argument 'smbclient', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/customisation.nix:69:16`
Means it's using a package alias, and should reference it by the canonical name instead? Like, `smbclient = samba; # added 2018-04-25` in `pkgs/top-level/aliases.nix:574` ?

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:2]:
```suggestion
, xdg_utils, gvfs, zip, unzip, gzip, bzip2, gnutar, p7zip, xz, imagemagick, darwin, libuchardet
, spdlog, xercesc, fmt, openssl, libssh, samba, neon, libnfs, libarchive }:
```

Also please do not unput darwin directly.

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:18]:
```suggestion
    ++ stdenv.lib.optional stdenv.isDarwin Cocoa # some macintosh stuff
```

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:19]:
```suggestion
    ++ [ openssl libssh samba /*aka smbclient*/ neon libnfs libarchive ]; # optional feature packages, like protocol support for Network panel, or archive formats
```

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:71]:
```suggestion
    description = "Linux port of FAR Manager v2";
```

Homepage is for links.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please squash the commits.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:71]: Homepage is the upstream project, which is a Linux port of the original product. Might not tell people much.
This one link is to the original product the Linux port is trying to mimic, which should be more widely known.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:2]: `darwin` is in prior code written by package maintainer. I won't dare change this. Especially that it's for Mac I which I can't test.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:18]: This is from existing code, I'd rather not change myself, even if syntax. I only added a comment because I wanted to add optional packages as a separate batch.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:19]: This was intentional, cmake wants `smbclient`, `smbclient` is an alias for `samba`. With a space it looks as if a few packages in the list are commented out. If it looks weird without a space, I'd just remove the comment, what do you think?

**Comment by hypersw** [comment]
> Please squash the commits.

I've squashed work in progress as much as I thought reasonable. The remaining commits are independent changes I'd rather see individually in the history if this were on my project; might not be the same reasoning for nixos, please advise. They're not about the update as it is but the small off-things I noticed during the update.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:2]: Wrapping I will do, just want to see the oborg checks complete before resetting by push.

File: pkgs/applications/misc/far2l/default.nix
Comment (thiagokokada) [inline pkgs/applications/misc/far2l/default.nix:74]: Please update the license to either `licenses.gpl2Only` or `licenses.gpl2Plus`: https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 108750` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>far2l</li>
  </ul>
</details>

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/misc/far2l/default.nix:72]: Not in the description. Link do not belong there.

**Comment by SuperSandro2000** [comment]
> The remaining commits are independent changes I'd rather see individually in the history if this were on my project; might not be the same reasoning for nixos, please advise. They're not about the update as it is but the small off-things I noticed during the update.

We usually just squash all those commits together to not explode the commit history.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: Okay.

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:19]:
```suggestion
    ++ lib.optional stdenv.isDarwin Cocoa
```

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:73]: Can you add yourself here as a maintainer?

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:2]:
```suggestion
  xdg-utils, gvfs, zip, unzip, gzip, bzip2, gnutar, p7zip, xz, imagemagick, Cocoa,
```

and inherit it from top level.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:73]: Should I? :) Merely noticed that the software is not workable on NixOS without GUI and contacted far2l folks and ruled out all I have to do is update. Maybe it's a build too bold to sign up as a maintainer. What is the policy?

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:19]: Are you 100% sure of this change? I mean, I haven't written this and I don't have a Mac to check it. I merely formatted the lines when adding new optional packages (to make more built-in archiver plugins work correctly) with a few comments on what is what. I can make this change, just a bit afraid to break something for people out there on Mac without even knowing. Just to confirm.

File: pkgs/applications/misc/far2l/default.nix
Comment (thiagokokada) [reply pkgs/applications/misc/far2l/default.nix:73]: > What is the policy?

AFAIK, there is no policy. Are you interested in seeing updates in this package and have some context on how the package works? You can add yourself as a maintainer.

The only thing it really changes is that you will receive a notification every time the package is update by someone that is not you, so you can see the changes and review them.

Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: OK then )

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/misc/far2l/default.nix:19]: > Are you 100% sure of this change?

Yes, this is a no-op change and just some cleanup. I can build it afterwards to check if it is broken.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: Looks like it's optional? What would be the correct way to request an optional package here?
`Cocoa ? []` ?

Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: Works for me. Ready for merge if so.

Comment (SuperSandro2000) [reply pkgs/applications/misc/far2l/default.nix]: > Looks like it's optional? What would be the correct way to request an optional package here?
Cocoa ? [] ?

```
buildInputs = lib.optionals stdenv.isDarwin [ Cocoa ];
```

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:3]:
```suggestion
, xdg-utils, gvfs, zip, unzip, gzip, bzip2, gnutar, p7zip, xz, imagemagick, Cocoa
, libuchardet, spdlog, xercesc, fmt, openssl, libssh, samba, neon, libnfs, libarchive }:
```

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, fetchpatch, makeWrapper, cmake, pkg-config, wxGTK30, glib, pcre, m4, bash
```

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:73]:
```suggestion
    maintainers = with maintainers; [ volth hypersw ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
3 little formatting changes after which it is good to be merged. I am providing a patch for darwin if anything fails.

Edit: Darwin does not build because the code base seems to be not compatible right now.

If you want you can remove the darwin specific code and mark it broken for darwin or just mark it broken.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: Nae, I mean in function arguments. _Anonymous function called without required argument 'Cocoa'_. See current code in the patch.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix:73]: )) were thinking about doing just that.

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: I can't change `Cocoa?[]` to `Cocoa` because it won't build on my system then. _Anonymous function called without required argument 'Cocoa'._ I guess you gotta have darwin for that to work.

I have no idea how codebase is supposed to compile against darwin. I suppose that means MacOS? I don't have a Mac to test it, so I'm locked out on this. For now the main goal for me is to make far2l runnable for regular GUI-less NixOS at least. So, speaking of darwin, probably leaving it exactly as it were before for now, or marking as broken (how specifically?) would be the best option.

Comment (SuperSandro2000) [reply pkgs/applications/misc/far2l/default.nix]: > I can't change Cocoa?[] to Cocoa because it won't build on my system then. Anonymous function called without required argument 'Cocoa'. I guess you gotta have darwin for that to work.

You need to specify it like 

``` 
748   aminal = callPackage ../applications/terminal-emulators/aminal {
749   | inherit (darwin.apple_sdk.frameworks) Carbon Cocoa Kernel;                                                                                                                          750   };
```

in top-level.

> I have no idea how codebase is supposed to compile against darwin. 

I tried building it against darwin and it is broken. I would suggest to remove all traces of Cocoa and add my next suggestion and then we can finally merge.

File: pkgs/applications/misc/far2l/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/far2l/default.nix:74]:
```suggestion
    platforms = platforms.all;
    # fails to compile with:
    # error: no member named 'st_ctim' in 'stat'
    broken = stdenv.isDarwin;
```

File: pkgs/applications/misc/far2l/default.nix
Comment (hypersw) [reply pkgs/applications/misc/far2l/default.nix]: OK.

Based on what I know about Nix, it should be possible to spawn a nix-shell environment which mocks itself as a proper Darwin and try building in that mode on my machine, is this so?

**Comment by hypersw** [comment]
@SuperSandro2000, thanks for your help!

**Pull request merged**
